engine/src/main/java/org/hibernate/search/engine/mapper/mapping/building/impl/IndexSchemaFilter.java

```diff
@@ -196,24 +199,111 @@ private String getPathFromSameIndexedEmbeddedSinceNoCompositionLimits(IndexedEmb
+		// we will use these lists for step 2, see below:
+		List<String> paths = new ArrayList<>();
+		List<IndexSchemaFilter> nodes = new ArrayList<>();
+
+		String shortestCycle = null;
+		String longestCycle = null;
+		IndexSchemaFilter longestCycleNode = null;
+		IndexSchemaFilter localParent = this;
+		StringBuilder path = new StringBuilder();
+		while ( localParent != null && localParent.definition != null ) {
+
+			path.insert( 0, localParent.definition.relativePrefix() );
+			paths.add( 0, path.toString() );
+			nodes.add( 0, localParent );
+
+			if ( localParent.definition.equals( definitionToBeAdded ) ) {
+				longestCycle = path.toString();
+				longestCycleNode = localParent;
+				if ( shortestCycle == null ) {
+					// set the first found cycle and keep it so till the end of the loop
+					shortestCycle = path.toString();
+				}
 			}
+
+			localParent = localParent.parent;
 		}
-		else {
-			/*
-			 * No recursion limits, no parent: this is the root.
-			 * I we reach this point, it means there was no recursion limit at all,
-			 * but we did not encounter the IndexedEmbedded we were looking for.
-			 */
+
+		if ( longestCycle == null ) {
 			return null;
 		}
+
+		// We found the longest cycle now let's see if it is included in any path:
+		// 1. First let's check if any node before the one that caused a cycle is going to break that cycle.
+		// To check the node where we've encountered the longest cycle we will first check the node itself, and then
+		// go to its parents till the root prepending the corresponding relative node path.
+		if ( isPotentiallyExcludedPath(
+				// when we check the node we don't need the node's relative path at the beginning of the path we are checking:
+				longestCycle.substring( longestCycleNode.definition.relativePrefix().length() ),
+				longestCycleNode
+		) ) {
+			return null;
+		}
+
+		// 2. maybe there's a node following the one causing the cycle that would break it at some point.
+		// If so we want to go one node at a time, and only check the "new nodes" i.e. if we encounter a same node for the second time
+		// it means even if that node at the current step may break the cycle - we've already checked it before and it didn't
+		// in the previous steps, meaning that if that node breaks a cycle, it wouldn't be the one we are currently in:
+		Set<IndexedEmbeddedDefinition> encounteredNodesSoFar = new HashSet<>();
+		// we don't want to start from the beginning but from the next node from the one at which we've found the longest cycle:
+		for ( int index = paths.indexOf( longestCycle ); index < nodes.size(); index++ ) {
+			IndexSchemaFilter node = nodes.get( index );
+			// so we know that we are in the cycle, and we know what the cycle looks like, if the current node can break a cycle
+			// it would mean that if we make a cycle permutation that starts with the "current node ^" and this same node can
+			// potentially break the cycle -- then we are ok to break it:
```

Comment:
I must admit this seems more complex than what I imagined, but maybe I'm missing some context.
Maybe let's discuss this over video chat.